### PR TITLE
Fix watches in e2e tests

### PR DESCRIPTION
**Pull request by tnozicka**
**What type of PR is this?**
/kind bug
/kind flake

**What this PR does / why we need it**:
Watches in e2e need to be re-estabished to survive closed calls

**Special notes for your reviewer**:
Split from https://github.com/kubernetes/kubernetes/pull/50102

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by tnozicka** [comment]
@kubernetes/sig-apps-pr-reviews @kubernetes/sig-node-pr-reviews

**Comment by kow3ns** [comment]
@tnozicka what milestone are you targeting?

**Comment by tnozicka** [comment]
@kow3ns 1.14 would be nice

**Comment by tnozicka** [comment]
@kow3ns or @janetkuo ptal at the sig-apps owned tests

@kubernetes/sig-node-pr-reviews for node related ones

**Review by kow3ns — APPROVED** [review]

**Comment by tnozicka** [comment]
@derekwaynecarr you are at the intersection of the 3 remaining OWNERS files, would you mind taking a look?

**Comment by alejandrox1** [comment]
Thank you for working on this @tnozicka !

/assign @Random-Liu @tallclair @ConnorDoyle 
for sig-node review

/assign @timothysc @andrewsykim 
for common and framework
/milestone v1.17

File: test/e2e/common/init_container.go [hunk 1]
```diff
@@ -90,14 +99,23 @@ var _ = framework.KubeDescribe("InitContainer [NodeConformance]", func() {
...
+				options.FieldSelector = fieldSelector
+				return podClient.Watch(options)
+			},
+		}
+		var events []watch.Event
 		ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), framework.PodStartTimeout)
 		defer cancel()
-		event, err := watchtools.UntilWithoutRetry(ctx, wr, conditions.PodCompleted)
+		event, err := watchtools.Until(ctx, startedPod.ResourceVersion, w,
+			recordEvents(events, conditions.PodCompleted),
```
Comment (timothysc) [inline test/e2e/common/init_container.go]: So I know this test was grandfathered in, but actually violates a bunch of requirements we listed here - 

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements 

namely on conditions and events.

**Review by timothysc — CHANGES_REQUESTED** [review]
So I like the improvement, but this specific test has a some issues that if we are going to refactor we should address - https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements

**Comment by ttousai** [comment]
Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on Nov. 18. Is this PR still intended for 1.17?

**Comment by ttousai** [comment]
***Correction:*** Code freeze is on Thursday, November 14.

Sorry for the mix up.

**Comment by tnozicka** [comment]
I'd prefer if this PR could stay limited to fixing the machinery. If we see the tests is, and was, not matching the criteria for conformance - Would removing it from conformance in this PR help? We could follow up with an issue for the owners to bring it back when it meats the criteria.

**Comment by tnozicka** [comment]
> Is this PR still intended for 1.17?
I think we can sort it out in time for code freeze

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Comment by tnozicka** [comment]
Had to rebase to fix conflicts with https://github.com/kubernetes/kubernetes/pull/85149 moving functions around and others that went in

**Comment by tnozicka** [comment]
@timothysc thanks for reviewing the PR. 

I've created an [issue](https://github.com/kubernetes/kubernetes/pull/66921) for the conformance test which doesn't meet the requirements.

Also can you please re-tag the PR as I was forced to rebase due to conflict?

**Comment by josiahbjorgaard** [comment]
/milestone clear

(code freeze v1.17)

**Comment by spiffxp** [comment]
needs rebase, ping me when you've done so

**Comment by spiffxp** [comment]
ping @tnozicka, this still needs a rebase; I'm willing to carry this PR forward if you don't have the time

File: test/e2e/common/init_container.go (hunk 1 above)
Comment (spiffxp) [reply test/e2e/common/init_container.go:223]: This came up during a recent conformance meeting in the context of a different test.  This reads like it violates those constraints, but those are really referring to the v1.Event resource and Condition field.  In this case, watch.Event is being used, which does have some guarantees about delivery, and conditions.PodCompleted is a function which looks at Pod.Status.Phase which is acceptable.

**Comment by spiffxp** [comment]
/lgtm
Thank you for keeping up with this!

**Comment by tnozicka** [comment]
@spiffxp thanks for the review and apologies for the delay

**Pull request merged**